webapp/src/main/java/com/box/l10n/mojito/service/tm/TMService.java

```diff
@@ -961,15 +980,29 @@ public String generateLocalized(
 
         logger.debug("Configuring pipeline for localized XLIFF generation");
 
-        BasePipelineStep translateStep = (BasePipelineStep) new TranslateStep(asset, repositoryLocale, inheritanceMode, status);
-        return generateLocalizedBase(asset, content, filterConfigIdOverride, filterOptions, translateStep, bcp47Tag);
+        boolean saveUsedTmTextUnitVariantIds = pullRunName != null;
+        TranslateStep translateStep = new TranslateStep(asset, repositoryLocale, inheritanceMode, status, saveUsedTmTextUnitVariantIds);
+        String generateLocalizedBase = generateLocalizedBase(asset, content, filterConfigIdOverride, filterOptions, translateStep, bcp47Tag);
+
+        if (saveUsedTmTextUnitVariantIds) {
+            saveUsedTmTextUnitVariantIds(asset, pullRunName, repositoryLocale.getLocale(), translateStep.getUsedTmTextUnitVariantIds());
+        }
+
+        return generateLocalizedBase;
+    }
+
+    void saveUsedTmTextUnitVariantIds(Asset asset, String pullRunName, Locale locale, List<Long> usedTmTextUnitVariantIds) {
+        PullRun pullRun = pullRunService.getOrCreate(pullRunName, asset.getRepository());
+        PullRunAsset pullRunAsset = pullRunAssetService.getOrCreate(pullRun, asset);
+        List<Long> uniqueUsedTmTextUnitVariantIds = usedTmTextUnitVariantIds.stream().distinct().collect(Collectors.toList());
```

Comment:
@garionpin Not sure if we should record exactly what's used to generate the file (ie potential duplicates, will need to change the UK again then) or if we record distinct entries (current implementation).

Reply:
Is this comment still applicable, now that you've added locale_id in the table?

Reply:
PS: having the same data recorded multiple times for a locale doesn't provide us with any additional information - I'd advocate for always uploading distinct().

Reply:
```
yes=it could happen

yes=it could happen
```